### ISSUE TITLE
Add filter bar in browsing view

### ISSUE THREAD

**Issue by SenorSmartyPants** [SenorSmartyPants/jellyfin-nmt#3]
Allow filtering current view.

**Comment by wienbarg** [comment]
I borrowed some code from Aeon-skin to use the keypad of the remote "SMS-style";-)
2=A, 22=B, 222=C and so on. Works for me.
`<a href="browse.php?CollectionType=tvshows&Name=A&Title=A" tvid="2"/>
...`
What do you think?

**Comment by SenorSmartyPants** [comment]
That's a possibility.

Index pages currently support the Home button to go back to index.php and INFO button to go to the categories.php which is also linked to from the index.php

**Comment by wienbarg** [comment]
categories.php is great!
But jumping/filtering directly in browse.php is faster. My wife likes because she is used to it from our "old" YAMJ ;-)

**Comment by SenorSmartyPants** [comment]
Understandable.

If I find the time/get in the right mood I'll take a look.

**Comment by wienbarg** [comment]
My wife just found out that my added lines collide with "your" way to navigate by numbers :-(
i.e. if she wants to filter by the letter "B" she has to press "2" once. But that opens the second entry in browse.php.
:-( I take care of that later...

**Comment by SenorSmartyPants** [comment]
I pushed an updated that includes SMS style speed dial TVIDs. 

- Title Letter (1-4 digits)
- Genre (3 letters)
- Year 4 digit code

Genres could collide, for example Family and Fantasy both resolve to 326. The first in alphabetically order is where you will end up. Genres could also collide with single letter codes. If this is a big deal I may try to implement collision avoidance.

YAMJ originally also had codes for library sections like TV Shows and Movies (the Other Section).  This has not been implemented.

**Comment by wienbarg** [comment]
Wow. Great. 👍🏻😁
Here are my first impressions (today's the first day we tried the new version):
I appreciate that you try to get the values for navbar(?) from jellyfin. And I think I know why you did it this way. But... 
Pageloading is soooo slow now 😒 Several seconds everytime. Is it due to the additional code?
And if I may be honest: I don't think that anybody can memorize all the additional codes. The letters can be found on the remote control but all the different genres. I don't know... Plus if I'm browsing tv-series and use the new method to jump to a letter the search result also includes movies. 
I am pretty happy with these lines in page.php:
```
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=%23&Title=%23" tvid="1"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=A&Title=A" tvid="2"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=B&Title=B" tvid="22"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=C&Title=C" tvid="222"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=D&Title=D" tvid="3"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=E&Title=E" tvid="33"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=F&Title=F" tvid="333"><a/>
…
…
…
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=Y&Title=Y" tvid="999"><a/>
<a href="browse.php?CollectionType=<?php echo $_GET["CollectionType"];?>&Name=Z&Title=Z" tvid="9999"><a/>
```
It's actually all I want. When looking for genres or production years I just jump to categories.php. There all genres and years are listed and I don't have to know the code.

**Comment by wienbarg** [comment]
btw I moved the code from browser.php to page.php. Now one-digit-letters (A, D, G...) work ;-)

**Comment by SenorSmartyPants** [comment]
> btw I moved the code from browser.php to page.php. Now one-digit-letters (A, D, G...) work ;-)

Single letter codes worked fine for me where they are.

I didn't really test it out on the PCH after I added the genres, etc. Yes, it is slow. I can trim that down to just the letters, or make them configurable.

I've been messing with creating a popup menu based on YAMJ skin PPP3.1 code. It works on my PC testing, but is also probably way too slow as well. Good thing I find performance tuning kind of stuff interesting.

**Comment by SenorSmartyPants** [comment]
I made a branch for some quick and dirty fixes for the slow page loads.

SMS codes only for Title letters now. They should stay filtered to tv show or movies based on what you were viewing.

Here's the direct ZIP download. Or checkout [QuickFilterFix](/SenorSmartyPants/jellyfin-nmt/tree/QuickFilterFix)

https://github.com/SenorSmartyPants/jellyfin-nmt/archive/QuickFilterFix.zip

**Comment by wienbarg** [comment]
> I made a branch for some quick and dirty fixes for the slow page loads.

> If you're curious, it was just getting filters for movies that was really slow. TV shows was an order of magnitude faster.

YES. Excellent. I only tested on PC -already much faster! PCH comes later.

**Comment by wienbarg** [comment]
> Single digit codes worked fine for me where they are. But that could be a minor difference between the 100s and the 200s PCHs. I am still writing out the old TVIDs so there are 2 copys of TVID="2"

In my first attempt I added the code at the top of browse.php. It did work except single digit codes. Because „your“ code was there too but several lines down.☝🏻
Now "my" code sits in page.php and in the resulting HTML-code it's not at the beginning of the page but in the <body> right after the first <table>.
But with your QuickFilterFix all this isn't necessary anymore!

**Comment by wienbarg** [comment]
> I've been messing with creating a popup menu based on YAMJ skin PPP3.1 code. It works on my PC testing, but is also probably way too slow as well. Good thing I find performance tuning kind of stuff interesting.

Sounds interesting. Is this to replace categories.php?
It's too bad, most of the screenshots on http://www.networkedmediatank.com/ are no longer available.
I found a youtube-video: https://youtu.be/TyTfiQi4g_Q?t=242 . Do you want to recreate that?

**Comment by SenorSmartyPants** [comment]
Yes, that's the menu. I have it mostly working, but performance is not great. And not respecting what you are viewing. Like showing movies after you made a selection from a tv show view.

**Comment by wienbarg** [comment]
There is the variable "CollectionType". I used it for my attempt on SMS-style filtering. It should be usable because of some "includes".
But it has to be fast... In the old YAMJ-skins everything was hardcoded so no requests to any server...

**Comment by SenorSmartyPants** [comment]
Do you have any JF libraries that aren't TV or movies?

**Comment by wienbarg** [comment]
> YAMJ originally also had codes for library sections like TV Shows and Movies (the Other Section). This has not been implemented.

Not necessary IMHO. If there's a shortcut to get to index.php you can switch to another library there.

> If you're curious, it was just getting filters for movies that was really slow. TV shows was an order of magnitude faster.

Aaaah, yes. Did some more testing today: There is a huge difference between TV shows and movies. I've lost track: where is the filtering done, listings.php? Isn't filtering TV shows the same as for movies?

**Comment by SenorSmartyPants** [comment]
The HOME button will take you to index.php

It was just getting the list of available filters for movies that was slow. Not the filtering itself.

**Comment by wienbarg** [comment]
> Do you have any JF libraries that aren't TV or movies?

My Jellyfin is as new as my JF-NMT 😉 I tried a music library for a short time but I was not very happy with the result.
For music I still use [http://www.networkedmediatank.com/showthread.php?tid=24832](http://www.networkedmediatank.com/showthread.php?tid=24832) The developer "disappeared" a long time ago but it's still working.

**Comment by wienbarg** [comment]
> The HOME button will take you to index.php

I know that! 😉

> It was just getting the list of available filters for movies that was slow. Not the filtering itself.

For better speed do it only in categories.php (or in the PPP-PopUp)🤓

**Comment by SenorSmartyPants** [comment]
Yep. Currently only categories.php is loading it.

Working on performance for pop up.

**Comment by wienbarg** [comment]
> Yep. Currently only categories.php is loading it.

For me filtering movies is still slower than with TV shows😢
Maybe I am using the wrong branch **again**? 😉😜
Which branch is the most current then?

**Comment by SenorSmartyPants** [comment]
I've merged the quick fix branch into master. So that's the most current.

I haven't done any performance testing on the actual filtering of movies vs tv shows. But all that is handled by JF and is out of my control.

**Comment by wienbarg** [comment]
> I've merged the quick fix branch into master. So that's the most current.

👍🏻😊

> filtering of movies […] But all that is handled by JF and is out of my control.

Interesting. Because there IS a difference when I use SMS codes for titles. But if it's JF there's nothing you can do.

**Comment by SenorSmartyPants** [comment]
Please describe where you are seeing a difference.
What page are you on, what is displayed, what are you clicking etc

**Comment by wienbarg** [comment]
I'll take some screenshots tomorrow👍🏻

**Comment by wienbarg** [comment]
First off all I have two Libraries that are relevant here: "TV Serien" (TV shows) and "Filme" (movies).
TV shows: http://192.168.2.89/nmt-210228/browse.php?Name=TV+Serien&parentId=fcc7edaa068ebdee843a2b5834b2d651&FolderType=CollectionFolder&CollectionType=tvshows
Movies: http://192.168.2.89/nmt-210228/browse.php?Name=Filme&parentId=7a2175bccb1f1a94152cbd2b2bae8f6d&FolderType=CollectionFolder&CollectionType=movies

When displaying the library with TV shows I use "555" to get series starting with the letter "L"
The same goes for movies. "555" for movies starting with "L".

My two observations are
1)
When I'm in TV shows I almost immediately get the result: Shows starting with "L".
For movies I used a stopwatch. It takes between 12 and 14(!) seconds until the new page appears. I have no idea why that happens. 
2)
I found the jellyfin logfile. The filtering (or is it a "search"?) for movies appears in the log:
`[2021-02-28 21:14:22.264 +01:00] [WRN] HTTP Response 200 to "192.168.2.89". Time (slow): 0:00:11.0032769. "http://192.168.2.89:8096/emby/Users/e8379f2693c44382a21906ba4f6d6a88/Items/?Fields=Path,ChildCount&Limit=27&IncludeItemTypes=Movie&SortBy=SortName&NameStartsWith=N&Recursive=true"`

BUT there's no sign of the search/filtering for TV-shows! As I understand it it's the same code for movies and TV-shows! Maybe Jellyfin handles them differently?

**Comment by wienbarg** [comment]
btw on the PC the speed difference is also there.
I used
`http://192.168.2.89/nmt-210228/browse.php?Name=Movies+-+L&FolderType=CollectionFolder&CollectionType=movies&Title=L
`
and 
`http://192.168.2.89/nmt-210228/browse.php?Name=TV+Shows+-+L&FolderType=CollectionFolder&CollectionType=tvshows&Title=L`

**Comment by SenorSmartyPants** [comment]
That JF logfile call is for loading info on the movies themselves. Not the filters for the movies. If the tv info is coming up fast JF doesn't output a WRN(warning) log entry.

And yes, since that is a delay in JF, it will be there for PC as well. Nothing I can do about that. 

Do you have a lot of movies? I have about 200 but don't have a delay like that. What version of JF are you running? I'm currently still on 10.5.5. I have an issue with 10.6 that I am waiting for a fix before I upgrade.

For debugging, if you go to data.php and change this line
`function apiCall($path, $debug = false, $includeAPIKey = true)`
to this
`function apiCall($path, $debug = true, $includeAPIKey = true)`
the web pages will display a URL link for each call to JF that you can click on. You'll see the data, and can time the page load if you like.

Not sure if I need recursive for all data those calls.

**Comment by wienbarg** [comment]
JF 10.4 here. Seems to be the latest stable there is. I am reluctant to try a different version than a stable on 😊
I just checked. I have about 370 movies and over 400 tv shows... Is this a lot?🤦‍♂️🙈
I will dive into your code (If I may) and play around with movie filtering. Maybe I'll be lucky...
Thanks for the debugging function -could be helpful.

In the meantime I have questions about itemDetails.php. I will open another issue☝🏻

**Comment by SenorSmartyPants** [comment]
Your repo is behind, but JF 10.6.4 is the latest stable release. 

There is a 10.7 RC4 coming soon, but I wouldn't run that for your day to day.

You're not going to fix movie filtering speed by looking at my code, that's a JF issue.

**Comment by SenorSmartyPants** [comment]
At least not that part of the loading speed.

**Comment by SenorSmartyPants** [comment]
Released in v0.3.0 https://github.com/SenorSmartyPants/jellyfin-nmt/releases/tag/v0.3.0
